core/src/main/java/org/apache/calcite/rel/rules/PruneEmptyRules.java

```diff
@@ -400,7 +401,13 @@ private static boolean isEmpty(RelNode node) {
 
     public void onMatch(RelOptRuleCall call) {
       SingleRel single = call.rel(0);
-      call.transformTo(call.builder().push(single).empty().build());
+      RelNode emptyValues = call.builder().push(single).empty().build();
+      if (single instanceof Sort) {
+        emptyValues = emptyValues.copy(
+            emptyValues.getTraitSet().replace(((Sort) single).getCollation()),
+            Collections.emptyList());
+      }
```

Comment:
Not only sort, but all the operators can have the same issue.
We should update emptyValues to have the same traits with single. Just collation is not enough, should be all the traits.

Reply:
Maybe we should put the trait logic in RelBuilder#empty ?

Reply:
we can try. But if only empty() considers the trait logic, the other methods don't, that might look odd.

Reply:
Agree with @hsyuan . This trait replacement is a one-off thing just for PruneEmptyRule. I am not convinced it should be available in RelBuilder.

Reply:
Done. Trait propagation applied to all RemoveEmptySingleRule instances (AGGREGATE, FILTER, PROJECT, SORT) + SORT_FETCH_ZERO_INSTANCE.
What about the rest of PruneEmptyRules (MINUS, UNION, INTERSECT, JOIN_LEFT, JOIN_RIGHT)?

Reply:
Propagating all traits does not seem a good solution. Right now there are some tests failing because of the following scenario: we have an enumerable single rel + logical empty values, so we replace the whole thing with an empty values (which in our case is a LogicalValues, as returned by RelBuilder#empty). If we try to propagate all traits it will fail, because we would be trying to propagate the EnumerableConvention towards a LogicalValues. So what should be our strategy here? Should we propagate all traits except the convention? Am I missing something in the trait propagation process?

Reply:
Moreover, what happens with e.g. Distribution? If we replace a SingleRel + EmptyValues = EmptyValues. Should the resulting EmptyValues have a values' distribution, or the SingleRel's distribution? What makes more sense?

Reply:
I'm confused, EmptyValues has no data, so what's the meaning of collation and distribution on it ?

Reply:
Collation is required in order to solve this bug, since we are removing the Sort:
Sort + EmptyValues => EmptyValues
if the resulting empty values does not have the sort's collation, we get a CannotPlanException. See https://issues.apache.org/jira/browse/CALCITE-3926 for more details.

Reply:
> Therefore, in order to fix this problem it seems clear that we need to propagate the collation into the EmptyValues.
Not really from my side, to me, it seems that the traits check should be tweaked for EmptyValues, it should satisfy all the required traits, the EmptyValues itself does nothing wrong.

Reply:
Thanks an interesting approach @danny0405 . I guess we could tweak EmptyValues to satisfy any collation... will work on that

Reply:
Unfortunately, I cannot find a clean way of tweaking EmptyValues' collation to satisfy any collation. Since propagating all traits into EmptyValues did not work either, I am going back to the original solution of propagating just collation into EmptyValues if we are removing a Sort from the plan.